docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,332 @@
+
+* Any reused data (for example, large sets of reference values or forms inputs) is better stored in a dedicated directory. Consider a descriptive folder name (for instance, `data`).
+
+With all the suggestions applied, your project's file structure might look like this:
+
+```sh
+.
+├── .testcaferc.json
+└── tests
+    ├── |- test_group1/
+    │   └── |-test1.js
+    │       |-test2.js
+    ├── |- test_group2/
+    │   └── |-test1.js
+    │       |-test2.js
+    ├── |- page_model/
+    │   └── |- page1.js
+    │       |- page2.js
+    ├── |- helpers/
+    │   └── |- helper1.js
+    │       |- helper2.js
+    └── |-data
+```
+
+## Setup and Teardown
+
+State management is an integral and important part of web testing. When your tests run, there are inevitably leftovers- database or local storage records, cache, or cookies. This data requires deletion during teardown. Extra steps your app requires to run (like adding records to the database) are performed during setup.
+
+One way to deal with this is to place the setup actions in the preceding test's `after` and `afterEach` hooks. While good for cleanup, these hooks create mutual dependence between your tests when used to set up for the following test. The success of a test is then influenced by a preceding one, which is not desirable.
+
+Consider a test that downloads a file and saves it to a folder. Since the file system persists between tests, this file requires deletion when the test is complete. This is better done with an `afterEach` hook:
+
+```js
+fixture `My fixture`
+    .page `http://example.com`
+    .afterEach( async t => {
+        await cleanDir();
+    });
+
+test('My test', async t => {
+    //test actions
+});
+```
+
+While good for cleanup, `after` and `afterEach` hooks create mutual dependence between tests when used to set up for the following test. The success rate of a test is then influenced by a preceding one, which is not desirable. Such tests need to run in a specific order and can't run in parallel.
+
+Use `before` or `beforeEach` to fullfill your test's prerequisites (for example to create a file necessary for a successful test run):
+
+```js
+fixture `Another fixture`
+    .page `http://example.com`
+    .beforeEach( async t => {
+        await setupFileSystem();
+    });
+
+test('Another test', async t => {
+    //test actions
+});
+```
+
+If the `before`/`beforeEach` setup is unsuccessful, the corresponding test does not run, which saves you time. If `after`/`afterEach` setup yields an error, the following test runs and probably fails due to the unsuccessful setup.
+
+To fulfill your test's prerequisites, use the `before` and `beforeEach` hooks. Clean up with the `after` and `afterEach` hooks.
+
+Use the `test.before` and `test.after` hooks to set the state that an individual test requires. Use the `fixture.beforeEach` and `fixture.afterEach` to set common state that is required across the board.
```

Comment:
> across the board

Is it a common expression? It sounds good, but does everyone know what it means?

Reply:
I find this use pretty common.
https://www.google.com/search?q=across+the+board&oq=across+the+board&aqs=chrome..69i57.271j0j7&sourceid=chrome&ie=UTF-8#dw-siw

Can be replaced with (although it becomes repetitive):

> Use the `test.before` and `test.after` hooks to set a state that an individual test requires. Use the `fixture.beforeEach` and `fixture.afterEach` to set a common state that all the tests require.

Let me know what you think (or we can ask someone else).

Reply:
`across the board` is pretty clear, but I feel that the more formal and repetitive variant is a bit more descriptive and easier to understand. IMHO

Reply:
> `across the board` is pretty clear, but I feel that the more formal and repetitive variant is a bit more descriptive and easier to understand. IMHO

Makes sense.

Use the `test.before` and `test.after` hooks to set a state that an individual test requires. Use the `fixture.beforeEach` and `fixture.afterEach` to set a common state that **all the tests in the suite require**.